### DISCUSSION TITLE
How to move _app.js to layout.js

### DISCUSSION THREAD

**Discussion by IRediTOTO** [vercel/next.js#41837] [Help]
Hi, in my `_app.js` have many global context like `next-auth` provinder. Can you show me how to move them to `layout.js`?
I'm very exciting about new folder `app` of NextJS.

**Comment by icyJoseph** [comment]
As far as I understand, you gotta use 'use client' directives:

```js
"use client";

import { Provider } from "./provider";

export default function RootLayout({ children }) {
  return (
    <html>
      <head></head>
      <body>
        <Provider value={{ foo: "bar" }}>{children}</Provider>
      </body>
    </html>
  );
}
```

Given `provider.js`:

```js
import { createContext } from "react";

const context = createContext({ foo: "baz" });

const { Provider } = context;

export { context, Provider };
```

Then on `app/foo/page.js`:

```js
"use client";

import { useContext } from "react";
import { context } from "../provider";

const Page = () => {
  console.log(useContext(context));
  return <button>foo page</button>;
};

export default Page;
```

I'll keep on learning and investigating, please do drop any info you might have yourself.

**Reply by Innei** [reply]
Hi, how to fetch root data in server component, and pass data into a context, bypass to the children element.

I wrote this. But not working.

```tsx
// app/layout.tsx
const fetchRootData = async () => {
  const data: InitialDataType & { reason?: any } = await fetchInitialData()

  return data
}
export const RootContainer: FC = (props) => {
  const initData = use(fetchRootData())

  return (
// ...
    <InitialContextProvider value={initData}> // this is a context, inject root data, and this FC mark as 'use client'
      {props.children}
    </InitialContextProvider>
// ...
  )
}
export default RootContainer;
```

Got runtime error.

```
event - compiled successfully in 231 ms (445 modules)
Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check your code at index.tsx:19.
Error: Unsupported Server Component type: undefined
    at attemptResolveElement (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.0.0_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1250:31)
    at retryTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.0.0_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1820:47)
    at performWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.0.0_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1856:33)
    at eval (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.0.0_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1257:40)
    at scheduleWork (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.0.0_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:52:25)
    at pingTask (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.0.0_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1256:29)
    at ping (webpack-internal:///(sc_server)/./node_modules/.pnpm/next@13.0.0_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1269:40)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)


```

index.tsx:19: `    <InitialContextProvider value={initData}> // this is a context, inject root data`

So, it's not support client context???

**Reply by IRediTOTO** [reply]
But if we write like this, what will run on server? :\ Will we lost the optimize feature of sever components for whole tree?

**Reply by IRediTOTO** [reply]
> ```tsx
> InitialContextProvider 
> ```

@Innei  You must add `'use-client'` on the first line of layout.js

**Reply by icyJoseph** [reply]
> But if we write like this, what will run on server? :\ Will we lost the optimize feature of sever components for whole tree?

From what I got at the conference, the truth is halfway, components, or subtrees that use context are sort of client side, JS-dependent sub-trees that cannot be rendered on the server. Same with stateful components, that useEffect or useState.

**Reply by Innei** [reply]
> > But if we write like this, what will run on server? :\ Will we lost the optimize feature of sever components for whole tree?
> 
> From what I got at the conference, the truth is halfway, components, or subtrees that use context are sort of client side, JS-dependent sub-trees that cannot be rendered on the server. Same with stateful components, that useEffect or useState.

No, I can't, if add 'use client', I can't fetch data in server side, and got request headers.

**Reply by Innei** [reply]
> > > But if we write like this, what will run on server? :\ Will we lost the optimize feature of sever components for whole tree?
> > 
> > 
> > From what I got at the conference, the truth is halfway, components, or subtrees that use context are sort of client side, JS-dependent sub-trees that cannot be rendered on the server. Same with stateful components, that useEffect or useState.
> 
> No, I can't, if add 'use client', I can't fetch data in server side, and got request headers.

I means, I want to fetch data in server side (and should set request headers based on `next/headers`), then inject data into context. So it only do in server side mode, cann't add 'use client', how do I do this.

**Reply by james-elicx** [reply]
> I means, I want to fetch data in server side (and should set request headers based on `next/headers`), then inject data into context. So it only do in server side mode, cann't add 'use client', how do I do this.

@Innei You can still fetch data on the server and pass it to a context on the client.

The context provider should have `'use client';` at the start of it, or you can put the context provider inside its own little wrapper component that has `'use client';` at the start. You can then fetch your data on the server as you normally would and pass this data to the client component as a prop.

The `use()` hook that you were using in the code snippet you posted should only be used in client components. In server components (ideally root layout is a server component), you can just `await` as you normally would.

**Reply by Innei** [reply]
> > I means, I want to fetch data in server side (and should set request headers based on `next/headers`), then inject data into context. So it only do in server side mode, cann't add 'use client', how do I do this.
> 
> @Innei You can still fetch data on the server and pass it to a context on the client.
> 
> The context provider should have `'use client';` at the start of it, or you can put the context provider inside its own little wrapper component that has `'use client';` at the start. You can then fetch your data on the server as you normally would and pass this data to the client component as a prop.
> 
> The `use()` hook that you were using in the code snippet you posted should only be used in client components. In server components (ideally root layout is a server component), you can just `await` as you normally would.

I'm sorry I didn't quite follow you. Is there a Demo? Thank you.
